Modules/mathmodule.c

```diff
@@ -1620,6 +1620,22 @@ completes the proof sketch.
 
 */
 
+
+/* Approximate square root of a large 64-bit integer.
+
+   Given `n` satisfying `2**62 <= n < 2**64`, return `a`
+   satisfying `(a - 1)**2 < n < (a + 1)**2. */
+
+uint64_t
+_approximate_isqrt(uint64_t n)
+{
+    uint32_t u = 1U + (n >> 62);
```

Comment:
`u = 1U + (uint32_t)(n >> 62)`?

Reply:
I removed the `uint32_t` casts on the basis that it makes the code look cleaner and the compiler will likely optimise them away anyway, given that the target type is `uint32_t`.

Reply:
Just double checked: at least on my machine (macOS 10.14.5, clang from the OS), with `-O3`, the generated assembly includes one `divb`, two `divl`s and one `divq`, so it looks as though the compiler's smart enough to optimise those divisions.

I have to confess I was a bit surprised by the `divb`.

And the second division could actually be done using 16-bit arithmetic. I don't know whether it's worth trying to persuade the compiler to do that. In _theory_ these division instructions could be a bottleneck, but in practise I suspect that all the overhead of the function call, argument passing, etc. will likely outweigh any advantage from optimising one division.

Reply:
Completely agree!